<a href="https://colab.research.google.com/github/JoshEliell/ColabFiles/blob/main/Actividad_Calificada_07_Datos_Temporales_274705.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos temporales**
### Limpieza y transformación de fechas, horas y duraciones en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: datos temporales |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Viajes de taxi en Nueva York, marzo de 2019 (NYC TLC, muestra publicada en un repositorio público de GitHub) |
| **Fecha de entrega** | [completar] |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** |Victor Joshua Hernandez Ramon |
| **Matrícula** |274705 |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en tus resultados (cita cifras).
7. Usa los métodos vistos en la sesión: `pd.to_datetime()`, el accesor `.dt`, `pd.Timedelta`, `.between()`, `tz_localize()`, `groupby()`, `sort_values()`, `.where()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Temporales_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset contiene 6,433 viajes de taxis amarillos y verdes de la ciudad de Nueva York registrados en **marzo de 2019**. Cada fila es un viaje. Las fechas y horas están en **hora local de Nueva York**.

| Columna | Descripción |
|---|---|
| `pickup` | Fecha y hora en que inicia el viaje |
| `dropoff` | Fecha y hora en que termina el viaje |
| `passengers` | Número de pasajeros |
| `distance` | Distancia recorrida, en **millas** |
| `fare` | Tarifa base (USD) |
| `tip`, `tolls` | Propina y peajes (USD) |
| `total` | Monto total cobrado (USD) |
| `color` | `yellow` o `green` |
| `payment` | Forma de pago |
| `pickup_zone`, `dropoff_zone` | Zona de origen y destino |
| `pickup_borough`, `dropoff_borough` | Distrito de origen y destino |

> **Dato útil**: en Nueva York, el horario de verano de 2019 comenzó el **domingo 10 de marzo**: a las 02:00 el reloj saltó a las 03:00.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Conversión y validación** | 20 | 18–20: conversión con formato explícito y verificación de `NaT`; viajes fuera de periodo detectados; análisis del cambio de horario correcto; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o conclusión. | 1–13: conversión sin verificar o validaciones incompletas; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Duraciones y anomalías** | 25 | 23–25: duración y velocidad calculadas correctamente; anomalías cuantificadas en todas las categorías; criterio de exclusión argumentado con evidencia. | 18–22: cálculos correctos; criterio parcial o con poca evidencia. | 1–17: cálculos con errores o anomalías no cuantificadas. | 0: no se entregó o es incorrecto. |
| **3. Componentes y patrones** | 25 | 23–25: componentes extraídos; viajes que cruzan la medianoche identificados; patrones por hora y día de la semana con `groupby()`; interpretación fundamentada. | 18–22: componentes y agrupaciones correctos; interpretación incompleta. | 1–17: agrupaciones con errores o sin interpretación. | 0: no se entregó o es incorrecto. |
| **4. Dataset limpio y comparación** | 20 | 18–20: dataset limpio construido con el criterio de 2.4 y verificado; comparación entre periodos con filtros temporales correcta; registro de decisión completo. | 14–17: dataset limpio correcto; comparación o registro incompletos. | 1–13: criterio aplicado de forma inconsistente o sin verificación. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Si el enlace no abre o requiere solicitar acceso, la actividad se considera no entregada.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [50]:
import pandas as pd

# Carga del dataset de taxis de Nueva York desde un repositorio público de GitHub.
# El archivo se lee con pandas; no se requiere ninguna otra biblioteca.
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/taxis.csv'
taxis = pd.read_csv(url)

print('Dimensiones:', taxis.shape)
taxis.head()

Dimensiones: (6433, 14)


,pickup,dropoff,passengers,distance,fare,tip,tolls,total,color,payment,pickup_zone,dropoff_zone,pickup_borough,dropoff_borough
0,2019-03-23 20:21:09,2019-03-23 20:27:24,1,1.60,7.0,2.15,0.0,12.95,yellow,credit card,Lenox Hill West,UN/Turtle Bay South,Manhattan,Manhattan
1,2019-03-04 16:11:55,2019-03-04 16:19:00,1,0.79,5.0,0.00,0.0,9.30,yellow,cash,Upper West Side South,Upper West Side South,Manhattan,Manhattan
2,2019-03-27 17:53:01,2019-03-27 18:00:25,1,1.37,7.5,2.36,0.0,14.16,yellow,credit card,Alphabet City,West Village,Manhattan,Manhattan
3,2019-03-10 01:23:59,2019-03-10 01:49:51,1,7.70,27.0,6.15,0.0,36.95,yellow,credit card,Hudson Sq,Yorkville West,Manhattan,Manhattan
4,2019-03-30 13:27:42,2019-03-30 13:37:14,3,2.16,9.0,1.10,0.0,13.40,yellow,credit card,Midtown East,Yorkville West,Manhattan,Manhattan


---
## Actividad 1. Conversión y validación (20 puntos)

### 1.1 Conversión (5 pts)

1. Muestra el tipo de dato de `pickup` y `dropoff`.
2. Convierte ambas columnas a tipo fecha con `pd.to_datetime()`, indicando el **formato explícito** que corresponde a los valores.
3. Verifica que la conversión no generó `NaT`: compara el número de faltantes antes y después.

In [51]:
# Escribe tu código aquí
# 1. Tipo de dato de pickup y dropoff
print("Tipos de dato originales:")
print(taxis[['pickup', 'dropoff']].dtypes)

# Valores faltantes antes de la conversion
faltantes_antes = taxis[['pickup', 'dropoff']].isna().sum()

# 2. Convertir a tipo datetime, formato (Ano-mes-dia Hora:minuto:segundo)
formato_fecha = '%Y-%m-%d %H:%M:%S'
taxis['pickup'] = pd.to_datetime(taxis['pickup'], format=formato_fecha)
taxis['dropoff'] = pd.to_datetime(taxis['dropoff'], format=formato_fecha)

# 3. Verificar NaT generados
faltantes_despues = taxis[['pickup', 'dropoff']].isna().sum()

print("\nFaltantes antes de la conversión:\n", faltantes_antes)
print("\nFaltantes despues de la conversión:\n", faltantes_despues)

Tipos de dato originales:
pickup     object
dropoff    object
dtype: object

Faltantes antes de la conversión:
 pickup     0
dropoff    0
dtype: int64

Faltantes despues de la conversión:
 pickup     0
dropoff    0
dtype: int64


### 1.2 Periodo de los datos (5 pts)

El dataset debería contener solo viajes de marzo de 2019.

1. Muestra la fecha mínima y máxima de `pickup` y de `dropoff`.
2. Cuenta y muestra los viajes cuyo `pickup` **no** está en marzo de 2019.
3. ¿Hay viajes con `pickup` en marzo y `dropoff` fuera de marzo? Cuéntalos.

In [52]:
# Escribe tu código aquí
# 1. Fecha minima y maxima
print("Pickup min:", taxis['pickup'].min(), "| Pickup max:", taxis['pickup'].max())
print("Dropoff min:", taxis['dropoff'].min(), "| Dropoff max:", taxis['dropoff'].max())

# 2. Viajes con pickup fuera de marzo de 2019
fuera_de_marzo = ~((taxis['pickup'].dt.year == 2019) & (taxis['pickup'].dt.month == 3))
print("\nViajes con pickup fuera de marzo 2019:", fuera_de_marzo.sum())

# 3. Viajes con pickup en marzo y dropoff fuera de marzo
en_marzo = (taxis['pickup'].dt.year == 2019) & (taxis['pickup'].dt.month == 3)
dropoff_fuera = ~((taxis['dropoff'].dt.year == 2019) & (taxis['dropoff'].dt.month == 3))
print("Viajes con pickup en marzo y dropoff fuera de marzo:", (en_marzo & dropoff_fuera).sum())

Pickup min: 2019-02-28 23:29:03 | Pickup max: 2019-03-31 23:43:45
Dropoff min: 2019-02-28 23:32:35 | Dropoff max: 2019-04-01 00:13:58

Viajes con pickup fuera de marzo 2019: 1
Viajes con pickup en marzo y dropoff fuera de marzo: 1


### 1.3 Cambio de horario (5 pts)

El 10 de marzo de 2019, en Nueva York, las horas entre 02:00 y 02:59 **no existieron**.

1. Filtra los viajes con `pickup` el 10 de marzo y cuenta cuántos hay en cada hora (`.dt.hour` y `groupby()` o `value_counts()`). ¿Hay viajes a las 2:00?
2. Localiza `pickup` en la zona `'America/New_York'` con `tz_localize()` usando `nonexistent='NaT'` y cuenta cuántos valores quedan como `NaT`.

In [53]:
# Escribe tu código aquí
# 1. Filtrar viajes del 10 de marzo de 2019 y contar por hora
fecha_cambio = pd.to_datetime('2019-03-10').date()
viajes_10_marzo = taxis[taxis['pickup'].dt.date == fecha_cambio]

# Contar los viajes agrupados por hora
viajes_por_hora = viajes_10_marzo.groupby(viajes_10_marzo['pickup'].dt.hour).size()
print("Viajes por hora el 10 de marzo de 2019:\n", viajes_por_hora)
print("¿Hay viajes a las 2:00?:", 2 in viajes_por_hora.index)

# 2. Localizar la zona horaria y contar NaT
pickup_tz = taxis['pickup'].dt.tz_localize('America/New_York', nonexistent='NaT')
print("\nValores NaT después de localizar pickup en America/New_York:", pickup_tz.isna().sum())

Viajes por hora el 10 de marzo de 2019:
 pickup
0     11
1      7
3      6
4      7
5      1
6      2
7      5
8      7
9      9
10    13
11    11
12    17
13     8
14    13
15     9
16    10
17    14
18     7
19     3
20     4
21     3
22    10
23     8
dtype: int64
¿Hay viajes a las 2:00?: False

Valores NaT después de localizar pickup en America/New_York: 0


### 1.4 Conclusión (5 pts)

¿Qué problemas temporales encontraste en esta actividad? ¿Qué indica el resultado de 1.3 sobre la zona horaria en que están registrados los datos? Responde con cifras.

**Respuesta 1.4:**

Problemas detectados: El principal problema encontrado es que el dataset contiene al menos un registro cuyo inicio (pickup) cae fuera del mes de análisis esperado. Además, existen viajes que inician el último día de marzo y cruzan a abril en el dropoff.

Sobre la zona horaria: El hecho de que no existan viajes a las 02:00 horas el dia 10 de marzo y que tz_localize devuelva 0 valores nulos (NaT), indica que los datos estan bien ajustados al horario local de Nueva York, el sistema respeto el salto del horario de verano omitiendo esa hora inexistente.

---
## Actividad 2. Duraciones y anomalías (25 puntos)

### 2.1 Duración de cada viaje (6 pts)

1. Crea la columna `duracion` = `dropoff` − `pickup`.
2. Crea `duracion_min` con la duración en **minutos** (usa `.dt.total_seconds()`).
3. Muestra el resumen estadístico de `duracion_min` con `describe()`.

In [54]:
# Escribe tu código aquí
# 1. Calcular la duracion
taxis['duracion'] = taxis['dropoff'] - taxis['pickup']

# 2. Calcular la duracion en minutos
taxis['duracion_min'] = taxis['duracion'].dt.total_seconds() / 60

# 3. Resumen estadistico
print("Resumen estadistico de duracion_min:")
print(taxis['duracion_min'].describe())

Resumen estadistico de duracion_min:
count    6433.000000
mean       14.349617
std        11.643892
min         0.000000
25%         6.500000
50%        10.900000
75%        18.516667
max       107.666667
Name: duracion_min, dtype: float64


### 2.2 Duraciones anómalas (6 pts)

Cuenta los viajes con:

- Duración **negativa** (termina antes de iniciar).
- Duración **cero**.
- Duración **menor a 1 minuto** (incluye los de duración cero).
- Duración **mayor a 3 horas**.

Muestra los viajes de duración cero con las columnas `pickup`, `dropoff`, `distance`, `fare` y `total`.

In [55]:
# Escribe tu código aquí
# Contar anomalías usando máscaras booleanas
print("Duracion negativa:", (taxis['duracion_min'] < 0).sum())
print("Duracion cero:", (taxis['duracion_min'] == 0).sum())
print("Duracion menor a 1 min (incluyendo cero):", (taxis['duracion_min'] < 1).sum())
print("Duracion mayor a 3 horas (> 180 min):", (taxis['duracion_min'] > 180).sum())

# Mostrar viajes con duracion cero
cols = ['pickup', 'dropoff', 'distance', 'fare', 'total']
viajes_cero = taxis[taxis['duracion_min'] == 0][cols]
print("\nDetalle de viajes con duración cero:\n", viajes_cero)

Duracion negativa: 0
Duracion cero: 6
Duracion menor a 1 min (incluyendo cero): 72
Duracion mayor a 3 horas (> 180 min): 0

Detalle de viajes con duración cero:
                   pickup             dropoff  distance  fare  total
1690 2019-03-22 06:24:14 2019-03-22 06:24:14       0.0  72.0   72.0
5493 2019-03-06 12:14:00 2019-03-06 12:14:00       0.0  10.0   10.0
5624 2019-03-11 14:04:50 2019-03-11 14:04:50       0.0   2.5    3.3
5638 2019-03-13 12:22:00 2019-03-13 12:22:00       0.0  10.0   10.0
6083 2019-03-19 15:34:00 2019-03-19 15:34:00       0.0   5.0    5.0
6356 2019-03-01 16:58:23 2019-03-01 16:58:23       0.0   3.0    4.8


### 2.3 Velocidad promedio (6 pts)

La velocidad promedio de un viaje, en millas por hora, es `distance / (duracion_min / 60)`.

1. Crea la columna `velocidad_mph` **solo para viajes con duración mayor que cero**; en los demás debe quedar como faltante (`NaN`). Sugerencia: calcula la velocidad para todos los viajes y conserva solo los valores válidos con `.where(condición)`, que deja `NaN` donde la condición es `False`.
2. Cuenta los viajes con velocidad mayor a 60 mph (velocidad poco plausible en tráfico urbano).
3. Muestra los 5 viajes con mayor velocidad, ordenados de mayor a menor, con `pickup`, `dropoff`, `duracion_min`, `distance`, `velocidad_mph` y `fare`.

In [56]:
# Escribe tu código aquí
# 1. Crear columna velocidad_mph (usando .where para dejar NaN si duracion <= 0)
velocidad_calculada = taxis['distance'] / (taxis['duracion_min'] / 60)
taxis['velocidad_mph'] = velocidad_calculada.where(taxis['duracion_min'] > 0)

# 2. Contar velocidades irreales
print("Viajes a más de 60 mph:", (taxis['velocidad_mph'] > 60).sum())

# 3. Top 5 viajes más rápidos
cols_vel = ['pickup', 'dropoff', 'duracion_min', 'distance', 'velocidad_mph', 'fare']
top_5_rapidos = taxis.sort_values(by='velocidad_mph', ascending=False)[cols_vel].head(5)
print("\nTop 5 viajes más rápidos:\n", top_5_rapidos)

Viajes a más de 60 mph: 11

Top 5 viajes más rápidos:
                   pickup             dropoff  duracion_min  distance  \
770  2019-03-02 03:16:59 2019-03-02 03:17:06      0.116667      9.40   
6089 2019-03-17 23:16:34 2019-03-17 23:16:52      0.300000     10.10   
491  2019-03-07 07:11:33 2019-03-07 07:11:39      0.100000      1.60   
5284 2019-03-22 06:24:02 2019-03-22 06:24:11      0.150000      1.70   
3241 2019-03-24 17:19:27 2019-03-24 17:19:29      0.033333      0.34   

      velocidad_mph  fare  
770     4834.285714   2.5  
6089    2020.000000  12.0  
491      960.000000   2.5  
5284     680.000000   2.5  
3241     612.000000  65.0  


### 2.4 Criterio de exclusión (7 pts)

Con base en 1.2, 2.2 y 2.3, define qué viajes excluirías del análisis y por qué. Considera:

a) ¿Todos los viajes de menos de 1 minuto son errores? Observa su distancia y tarifa.

b) ¿Qué umbrales (duración, velocidad, periodo) usarás? Justifícalos.

**Respuesta 2.4:**

a) No necesariamente todos los viajes menores a 1 minuto puede que sean errores de captura. Al revisar la distancia y la tarifa base (fare), algunos de estos viajes registran 0 millas pero tienen un cobro minimo, esto puede representar viajes cancelados tras abordar al pasajero.
b) Umbrales justificados: Evitaria tomar en cuenta cualquier viaje que cumpla alguna de las siguientes condiciones:   

duracion_min <= 0: Es imposible y produce division entre cero al calcular la velocidad.

velocidad_mph > 60: En un entorno urbano como Nueva York, una velocidad promedio superior a 60 millas por hora es muy improbable y puede ser un error en los datos de distancia.

pickup < '2019-03-01': Viajes que no pertenecen al periodo de estudio esperado.   

---
## Actividad 3. Componentes y patrones (25 puntos)

### 3.1 Extraer componentes (6 pts)

Crea las columnas:

- `hora`: hora de `pickup`.
- `dia_num`: día de la semana de `pickup` (lunes = 0).
- `dia_nombre`: nombre del día de `pickup`.
- `fin_semana`: `True` si es sábado o domingo.
- `fecha`: solo la fecha (sin hora) de `pickup`.
- `cruza_medianoche`: `True` si la **fecha** de `dropoff` es distinta de la de `pickup`.

Cuenta cuántos viajes cruzan la medianoche.

In [57]:
# Escribe tu código aquí
# Creacion de columnas
taxis['hora'] = taxis['pickup'].dt.hour
taxis['dia_num'] = taxis['pickup'].dt.dayofweek
taxis['dia_nombre'] = taxis['pickup'].dt.day_name()
taxis['fin_semana'] = taxis['dia_num'].isin([5, 6]) # 5 = Sábado, 6 = Domingo
taxis['fecha'] = taxis['pickup'].dt.date

# Viajes que cruzan la medianoche
taxis['cruza_medianoche'] = taxis['pickup'].dt.date != taxis['dropoff'].dt.date

# Suma
print("Total de viajes que cruzan la medianoche:", taxis['cruza_medianoche'].sum())

Total de viajes que cruzan la medianoche: 50


### 3.2 Patrón por hora del día (7 pts)

Con `groupby('hora')`, calcula para cada hora: el **número de viajes** y la **mediana** de `duracion_min`. Indica la hora con más viajes y la hora con menos.

Usa la mediana (no la media) porque la duración tiene valores extremos.

In [58]:
# Escribe tu código aquí
# Agrupacion por hora para contar viajes y calcular mediana de duracion
patron_hora = taxis.groupby('hora').agg(
    num_viajes=('duracion_min', 'count'),
    mediana_duracion=('duracion_min', 'median')
)

print(patron_hora)
print("\nHora con mas viajes:", patron_hora['num_viajes'].idxmax())
print("Hora con menos viajes:", patron_hora['num_viajes'].idxmin())

      num_viajes  mediana_duracion
hora                              
0            205         10.733333
1            110          9.633333
2            101          9.183333
3             67          9.583333
4             57         10.500000
5             51          8.700000
6            142          7.833333
7            221          9.583333
8            313         11.316667
9            319         11.616667
10           325         11.883333
11           295         12.033333
12           334         12.775000
13           316         10.558333
14           360         11.841667
15           330         12.058333
16           336         11.583333
17           388         11.975000
18           417         11.100000
19           406         10.650000
20           368          9.258333
21           355         11.083333
22           321          9.816667
23           296         10.575000

Hora con mas viajes: 18
Hora con menos viajes: 5


### 3.3 Patrón por día de la semana (6 pts)

Calcula por día de la semana el **número de viajes** y la **mediana** de `duracion_min` y de `fare`. Presenta los días en orden de lunes a domingo (sugerencia: agrupa por `['dia_num', 'dia_nombre']`).

Toma en cuenta que marzo de 2019 no tiene el mismo número de cada día de la semana: calcula también el **promedio de viajes por fecha**, dividiendo entre el número de fechas distintas de ese día de la semana (usa la columna `fecha` y `.nunique()`).

In [59]:
# Escribe tu código aquí
# Agrupar por dia (numerico y nombre para mantener el orden de lunes a domingo)
patron_dia = taxis.groupby(['dia_num', 'dia_nombre']).agg(
    num_viajes=('duracion_min', 'count'),
    mediana_duracion=('duracion_min', 'median'),
    mediana_fare=('fare', 'median'),
    num_fechas_distintas=('fecha', 'nunique')
)

# Promedio de viajes dividiendo el total de viajes entre la cantidad de fechas del dia
patron_dia['promedio_viajes_por_fecha'] = patron_dia['num_viajes'] / patron_dia['num_fechas_distintas']

print(patron_dia[['num_viajes', 'mediana_duracion', 'mediana_fare', 'promedio_viajes_por_fecha']])

                    num_viajes  mediana_duracion  mediana_fare  \
dia_num dia_nombre                                               
0       Monday             708         10.541667           9.5   
1       Tuesday            825         11.216667           9.5   
2       Wednesday          966         11.291667           9.5   
3       Thursday           905         11.766667           9.5   
4       Friday            1115         11.150000           9.5   
5       Saturday          1046         10.466667           9.0   
6       Sunday             868          9.983333           9.0   

                    promedio_viajes_por_fecha  
dia_num dia_nombre                             
0       Monday                         177.00  
1       Tuesday                        206.25  
2       Wednesday                      241.50  
3       Thursday                       181.00  
4       Friday                         223.00  
5       Saturday                       209.20  
6       Sunday       

### 3.4 Interpretación (6 pts)

a) ¿Qué patrones observas por hora y por día? ¿Coinciden la hora con más viajes y la hora con mayor duración mediana? ¿Qué podría explicarlo?

b) ¿Por qué fue necesario calcular el promedio de viajes por fecha en 3.3 en lugar de comparar solo el total por día de la semana?

**Respuesta 3.4:**

a) La hora con mas viajes son las tardes/noches, 18:00 - 19:00 hrs se puede pensar que es por la salida de los empleados, la hora con mayor duración mediana en cambio va hacia las 14:00 - 17:00 hrs que se puede penasr como la hora con mas trafico. No coinciden porque un alto volumen de viajes cortos en la noche puede explicar esto.

b) Fue necesario calcular el promedio de viajes por fecha porque marzo tiene 31 dias, lo que significa que algunos días de la semana aparecerán 5 veces y otros solo 4. Comparar sin esta normalizacion llevaria a conclusiones falsas como decir que los lunes tienen más demanda simplemente porque hubo un dia mas en el mes.

---
## Actividad 4. Dataset limpio y comparación (20 puntos)

### 4.1 Dataset limpio (6 pts)

Crea `taxis_limpio` aplicando el criterio que definiste en 2.4. Imprime cuántos viajes se excluyeron por cada condición y en total, y verifica que `taxis_limpio` cumple todas las condiciones.

In [60]:
# Escribe tu código aquí
# Mascara de exclusion
m_marzo = (taxis['pickup'].dt.year == 2019) & (taxis['pickup'].dt.month == 3)
m_duracion_valida = (taxis['duracion_min'] > 0)
m_velocidad_plausible = (taxis['velocidad_mph'] <= 60)

print("Excluidos por fecha (fuera de marzo):", (~m_marzo).sum())
print("Excluidos por duración (cero o negativa):", (~m_duracion_valida).sum())
print("Excluidos por velocidad (> 60 mph):", (~m_velocidad_plausible).sum())

# Mascara total
mascara_limpia = m_marzo & m_duracion_valida & m_velocidad_plausible

taxis_limpio = taxis[mascara_limpia].copy()
print("\nTotal de viajes excluidos en conjunto:", (~mascara_limpia).sum())
print("Dimensiones de taxis_limpio:", taxis_limpio.shape)

# Verificacion de que las anomalias se eliminaron
print("Velocidad maxima en limpio:", taxis_limpio['velocidad_mph'].max())
print("Duración minima en limpio:", taxis_limpio['duracion_min'].min())

Excluidos por fecha (fuera de marzo): 1
Excluidos por duración (cero o negativa): 6
Excluidos por velocidad (> 60 mph): 17

Total de viajes excluidos en conjunto: 18
Dimensiones de taxis_limpio: (6415, 23)
Velocidad maxima en limpio: 54.174757281553404
Duración minima en limpio: 0.05


### 4.2 Comparación entre periodos (6 pts)

Con `taxis_limpio`, compara los viajes que inician entre las **7:00 y las 9:59** en **días hábiles** contra los del **mismo horario en fin de semana**. Para cada grupo reporta: número de viajes, mediana de `duracion_min` y mediana de `velocidad_mph`.

In [61]:
# Escribe tu código aquí
# Filtrar 7:00 a 9:59
m_horario = taxis_limpio['hora'].between(7, 9)

# Separar dias hábiles y fines de semana
habiles = taxis_limpio[m_horario & ~taxis_limpio['fin_semana']]
fines = taxis_limpio[m_horario & taxis_limpio['fin_semana']]

print("Horario 7:00 a 9:59 en dias habiles")
print("Viajes totales:", len(habiles))
print("Mediana de duración:", round(habiles['duracion_min'].median(), 2), "minutos")
print("Mediana de velocidad:", round(habiles['velocidad_mph'].median(), 2), "mph")

print("\nHorario 7:00 a 9:59 en Fin de Semana")
print("Viajes totales:", len(fines))
print("Mediana de duración:", round(fines['duracion_min'].median(), 2), "minutos")
print("Mediana de velocidad:", round(fines['velocidad_mph'].median(), 2), "mph")

Horario 7:00 a 9:59 en dias habiles
Viajes totales: 687
Mediana de duración: 11.97 minutos
Mediana de velocidad: 9.09 mph

Horario 7:00 a 9:59 en Fin de Semana
Viajes totales: 163
Mediana de duración: 7.63 minutos
Mediana de velocidad: 13.47 mph


### 4.3 Registro de decisión (8 pts)

Redacta el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Conversión**: formato utilizado y verificación.
- **Zona horaria**: en qué zona están los datos y cómo lo verificaste.
- **Criterio de exclusión**: condiciones, umbrales y justificación.
- **Resultado**: viajes antes, después y excluidos.
- **Hallazgo de 4.2** y **limitaciones** del análisis.

**Respuesta 4.3:**

Conversión: Se forzo el formato especifico '%Y-%m-%d %H:%M:%S' usando pd.to_datetime(). Se verifico contabilizando nulos, comprobando que la conversion no introdujo valores NaT.   

Zona horaria: Al cruzar los datos contra el horario de verano, se confirmo que no hay viajes a las 02:00 am el 10 de marzo. Esto demuestra que los datos si estan ajustados correctamente al horario local de Nueva York y respetaron el salto de horario.

Criterio de exclusión: Se eliminaron registros previos o posteriores a marzo de 2019, duraciones iguales o menores a 0 minutos, y viajes con una velocidad promedio mayor a 60 mph, esto para eliminar datos que pueden considerarse incorrectos o errores del sistema.   

Resultado: De un total inicial de 6,433 viajes, se eliminaron 18 registros que presentaban anomalias, obteniendo un dataset limpio final de 6,415 viajes.  

Hallazgo de 4.2 y limitaciones: Los viajes matutinos en fin de semana muestran una duracion mediana menor y velocidades promedio mayores frente a los dias habiles, demostrando la ausencia de congestión vial laboral. La limitacion principal es ignorar los viajes atipicamente lentos que podrían ser atascos o accidentes.  

---
## Reflexión final (10 puntos)

1. ¿Qué ventajas tuvo indicar el formato explícito en 1.1 frente a dejar que pandas lo infiera? ¿En qué situación la inferencia habría producido errores silenciosos?
2. Supón que `pickup` se hubiera registrado en UTC y `dropoff` en hora local de Nueva York. ¿Qué efecto tendría en las duraciones calculadas? ¿Cómo lo habrías detectado con las herramientas de esta actividad?
3. Si se entrenara un modelo para predecir la duración de un viaje, ¿qué problemas causarían las anomalías temporales que encontraste si no se limpiaran?

**Respuesta final:**

1. Usar un formato explicito hace que el codigo sea más rapido y evita que Pandas confunda el día con el mes (por ejemplo, leer el 4 de marzo como el 3 de abril), lo cual meteria errores silenciosos en nuestro analisis.

2. Si el inicio fuera UTC y el final fuera hora local, los viajes parecerian tomar horas negativas (terminarian antes de empezar por la diferencia de zonas horarias). Nos habriamos dado cuenta en el paso 2.2 al contar las duraciones negativas.

3. Si no limpiamos los datos, el modelo aprenderia de información imposible, como teletransportarse con un viaje instantaneos o viajar a mas de 60 mph en la ciudad y terminaría prediciendo tiempos de llegada irreales o demasiado cortos para los usuarios.  

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.